src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.Unix.cs

```diff
@@ -0,0 +1,156 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Diagnostics;
+
+namespace System.Formats.Tar
+{
+    internal static partial class TarHelpers
+    {
+        private static readonly Lazy<UnixFileMode> s_umask = new Lazy<UnixFileMode>(DetermineUMask);
```

Comment:
Is it OK to cache this for the lifetime of the process?

Technically this could change overtime of the process, right? Someone else could modify the umask in between two calls to the System.Formats.Tar library.

Reply:
How common should be expect an umask to change? Can it change within the same user session? Is it possible to affect already running processes with an umask change, or would it only affect new processes?

It's a good point to raise though.

And I can understand why tmds did it this way - There are two places where we need to know the umask. But I wonder if instead of having a `Lazy`, we determine the umask on demand:

- When calling a `TarFile.Extract*` method: The umask could be determined once, at the beginning of calling any of these Extract methods. The umask value can then be passed to any `TarFile.ExtractToFile` internal call, when the entry is a directory.

- When the user extracts an individual entry: If the `TarEntry.ExtractToFile` method is _not_ being called from inside a `TarFile.Extract*` method, we can determine the umask too, if the entry is a directory.

Thoughts?

Reply:
The `umask` applies to the whole process. Any .NET code that changes the `umask` is not thread-safe.

We should not encourage users to change the umask before using these methods.

If there are use-cases for this, it would be preferable for the API to accept a mask argument so it can work in a thread-safe way.

Reply:
> We should not encourage users to change the umask before using these methods.

I'm not talking about encouraging people to change the umask of a process. I'm talking about what happens if they do.

Reply:
Currently, changing the `umask` later won't affect the cached value.

If we determine the `umask` each time the API gets called users may rely on that.

I think we should avoid that and let user create new issues if they this, so it can be provided through the proper API.